lib/Dialect/HW/HWModuleOpInterface.cpp

```diff
@@ -0,0 +1,89 @@
+//===- HWModuleOpInterface.cpp.h - Implement HWModuleLike  ------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file implements HWModuleLike related functionality.
+//
+//===----------------------------------------------------------------------===//
+
+#include "circt/Dialect/HW/HWOpInterfaces.h"
+#include "mlir/Transforms/DialectConversion.h"
+
+using namespace circt;
+using namespace hw;
+
+//===----------------------------------------------------------------------===//
+// HWModuleLike Signature Conversion
+//===----------------------------------------------------------------------===//
+
+static LogicalResult convertModuleOpTypes(HWModuleLike funcOp,
+                                          const TypeConverter &typeConverter,
+                                          ConversionPatternRewriter &rewriter) {
+  ModuleType type = funcOp.getHWModuleType();
+  if (!type)
+    return failure();
+
+  // Convert the original function types.
+  // Update the function signature in-place.
+  SmallVector<ModulePort> newPorts;
+  TypeConverter::SignatureConversion result(type.getNumInputs());
+  unsigned atInput = 0;
+  unsigned curInputs = 0;
+  for (auto &p : type.getPorts()) {
+    if (p.dir == ModulePort::Direction::Output) {
+      SmallVector<Type, 1> newResults;
+      if (failed(typeConverter.convertType(p.type, newResults)))
+        return failure();
+      for (auto np : newResults)
+        newPorts.push_back({p.name, np, p.dir});
+    } else {
+      if (failed(typeConverter.convertSignatureArg(
+              atInput++,
+              p.dir == ModulePort::Direction::Input ? p.type
```

Comment:
This is something we need a better solution for. Can you reverse this and specify the InOut direction to increase the clarity of what's going on here? Maybe a comment as well? This line tripped me up for a minute.